### PR TITLE
feat(docs-infra): set up Google Analytics 4 along with keeping legacy Universal Analytics

### PR DESCRIPTION
**Pull request by devversion**
We currently use Universal Analytics. This is deprecated in favor of
Google Analytics 4 and UA will stop processing hits in October 2023.

This change intends to prepare us for this migration, and to already
pre-populate our GA4 property (there is no way to migrate existing data
/properties into a GA4 property -- a new one needs to be created).

This will help us minimize the data gap so that we can:

* Continue to look at the UA property with the full time span until
  October 2023
* Can start using the GA4 property long-term in the future, starting
  with data even before Universal Analytics stops processing new data.

We need to keep the existing `analytics.js` setup. Initially we have
considered using `gtag.js` for both the UA and GA4 properties, as it
supports that, but that doesn't work with our strict trusted types
enforcement because it results in multiple `gtag.js` scripts (specific
versions for UA or GA4) that recreate the same trusted type policies.
This causes runtime errors and breaks the setup.

Instead, with continued use of `analytics.js` we have the benefit of
a good separation of trusted types + events and configuration. There is
some problematic with translation of Universal Analytics Events to GA4,
or the other way around (even though we don't use custom events
currently)

We also do not need to send page views for our GA4 property because GA4
with gtag supports this automatically (respecting the history state --
using the `Enhanced measurement events` setting in the UI).
For our UA legacy instance we continue to dispatch events manually. This
logic can be removed in the future.

More details can be found here:
https://docs.google.com/document/d/1aK8u4ZlXbqQ2wMqmgSX7Ces8iLgamC13oCoG6VeBruA/edit?usp=sharing&resourcekey=0-EVe-Rhnme3bj_pkz2RcOmw.

### CODE REVIEW

File: aio/angular.json [hunk 1]
```diff
@@ -137,7 +137,8 @@
           "builder": "@angular-devkit/build-angular:dev-server",
           "options": {
             "headers": {
-              "Content-Security-Policy": "require-trusted-types-for 'script'; trusted-types angular angular#bundler angular#unsafe-bypass aio#analytics google#safe"
+              // Keep in sync with the `firebase.json` file.
+              "Content-Security-Policy": "require-trusted-types-for 'script'; trusted-types angular angular#bundler angular#unsafe-bypass aio#analytics google#safe goog#html; report-uri https://csp.withgoogle.com/csp/angular.io"
```
Comment (gkalpak) [inline aio/angular.json:141]: Is `report-uri` needed here? Do we need to report violations in dev mode?

File: aio/src/app/shared/analytics.service.ts [hunk 2]
```diff
@@ -0,0 +1,93 @@
+import { Inject, Injectable } from '@angular/core';
+
+import { environment } from '../../environments/environment';
+import { WindowToken } from 'app/shared/window';
+import { unwrapScriptUrlForSink, scriptUrl } from 'safevalues';
```
Comment (gkalpak) [inline aio/src/app/shared/analytics.service.ts:5]: Nit: Order imported identifiers alphabetically 😇

```suggestion
import { scriptUrl, unwrapScriptUrlForSink} from 'safevalues';
```

File: aio/src/app/shared/analytics.service.ts [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+
+/** Extension of `Window` with potential Google Analytics fields. */
+interface WindowWithGtag extends Window {
+  dataLayer?: any[];
+  gtag?(...args: any[]): void;
+  ga?(...args: any[]): void;
```
Comment (gkalpak) [inline aio/src/app/shared/analytics.service.ts:11]: Does `ga` come from `gtag` or is it from the legacy analytics?

File: aio/src/app/shared/analytics.service.ts [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+};
+
+@Injectable()
+/**
+ * Google Analytics Service - captures app behaviors and sends them to Google Analytics.
+ *
+ * Note: Presupposes that the legacy `analytics.js` script has been loaded on the
+ * host web page.
+ *
+ * Associates data with properties determined the environment configurations:
```
Comment (gkalpak) [inline aio/src/app/shared/analytics.service.ts:21]:
```suggestion
 * Associates data with properties determined from the environment configurations:
```

File: aio/src/app/shared/analytics.service.ts [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+    // Won't re-send if the url hasn't changed.
+    if (url === this.previousUrl) { return; }
+    this.previousUrl = url;
+
+    this._legacyGa('set', 'page', '/' + url);
+    this._legacyGa('send', 'pageview');
+  }
+
+  private _gtag(...args: any[]) {
+    const gtagFn = (this.window as any).gtag;
```
Comment (gkalpak) [inline aio/src/app/shared/analytics.service.ts:54]: Is the `as any` cast necessary now that `this.window` is typed as `WindowWithGtag`?

File: aio/src/app/shared/analytics.service.ts [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+    if (gtagFn) {
+      gtagFn(...args);
+    }
+  }
+
+  private _legacyGa(...args: any[]) {
+    const gaFn = (this.window as any).ga;
```
Comment (gkalpak) [inline aio/src/app/shared/analytics.service.ts:61]: Is the `as any` cast necessary now that `this.window` is typed as `WindowWithGtag`?

File: aio/src/app/shared/analytics.service.ts [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+    }
+  }
+
+  private _installGlobalSiteTag() {
+    const window = this.window;
+    const url: TrustedScriptURL =
+      scriptUrl`https://www.googletagmanager.com/gtag/js?id=${environment.googleAnalyticsId}`;
+
+    // skip `gtag` for Protractor e2e tests.
+    if (window.name.includes('NG_DEFER_BOOTSTRAP')) {
```
Comment (gkalpak) [inline aio/src/app/shared/analytics.service.ts:73]: Will the window name still include `NG_DEFER_BOOTSTRAP` by the time this service is instantiated?
(I don't recall how Protractor works :grin:)

File: aio/src/app/shared/analytics.service.ts [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+      return;
+    }
+
+    // Note: This cannot be an arrow function as `gtag.js` expects an actual `Arguments`
+    // instance with e.g. `callee` to be set. Do not attempt to change this and keep this
+    // as much as possible in sync with the original tracking code snippet.
```
Comment (gkalpak) [inline aio/src/app/shared/analytics.service.ts:79]: What does the "original tracking code snippet" refer to? 😕

**Review by gkalpak — APPROVED** [review]
A couple of questions. Otherwise lgtm :+1:

_EDIT: I realized there are more than a couple of questions, but still all pretty minor_ 😛

File: aio/src/app/shared/analytics.service.ts (hunk 8 above)
Comment (devversion) [reply aio/src/app/shared/analytics.service.ts:79]: I clarified that with more comment lines

File: aio/src/app/shared/analytics.service.ts (hunk 7 above)
Comment (devversion) [reply aio/src/app/shared/analytics.service.ts:73]: I checked the source code and made threw error here (tests stopped working). Seems to work.

File: aio/src/app/shared/analytics.service.ts [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+    window.gtag = function() { window.dataLayer?.push(arguments); };
+    window.gtag('js', new Date());
+
+    // Configure properties before loading the script. This is necessary to avoid
+    // loading multiple instances of the gtag JS scripts.
+    window.gtag('config', environment.googleAnalyticsId);
+
+    // do not load the library when we are running in an e2e protractor test.
+    if (this.isProtractor) {
+      return;
```
Comment (gkalpak) [inline aio/src/app/shared/analytics.service.ts:100]: Could we move this higher up?

**Review by gkalpak — APPROVED** [review]
One minor comment.
Otherwise lgtm :+1:

**Review by dylhunn — APPROVED** [review]
reviewed-for: size-tracking

**Comment by jessicajaniuk** [comment]
@devversion This PR has conflicts with `14.0.x`. I've updated to `target: minor`. Please create a patch PR if you'd like to land in `14.0.x`.

**Pull request closed without merging**